exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableTemplate.java

```diff
@@ -661,16 +669,15 @@ public PutStatus put(int incomingRowIdx, IndexPointer htIdxHolder, int hashCode,
 
     // if startIdx is non-empty, follow the hash chain links until we find a matching
     // key or reach the end of the chain (and remember the last link there)
-    for ( currentIdxHolder.value = startIdx;
-          currentIdxHolder.value != EMPTY_SLOT;
-          /* isKeyMatch() below also advances the currentIdxHolder to the next link */) {
-
+    for ( int currentIndex = startIdx;
+          currentIndex != EMPTY_SLOT;
+          currentIndex = lastEntryBatch.nextLinkInHashChain(currentIndex)) {
```

Comment:
Presumably this function call gets inlined by the JVM ?  Since it gets called for every Put operation, this is a performance critical code.  Also, the comment in the method says 'this method should only be called following a False from isKeyMatch()'.  It would have been better to call this in the 'else' portion after the 'if' check on line 679 even though the loop condition is being checked after a False is returned (but it is not intuitive).

Reply:
Given that this method is so simple and short, and being called a lot - it is the ultimate candidate for JVM inlining.
I initially tried to build an iterator for this use, but it gets very complicated as the next links jump from one BatchHolder to another; still the for() loop tries to show a classic traversal of a linked chain; moving the "next" method inside the body spoils this "iterator like" logic.
    The comment should have also said "never call this when currentIndex == EMPTY_SLOT", which is trivial when used in the for().
    I reworded that comment to reflect the "iterator like" use.